portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,9 +33,73 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    private static final Gson gson = new Gson();
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Query query = new Query("Task").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Task> tasks = new ArrayList<>();
+        int displayTasks = Integer.parseInt(getParameter(request, "quantity", ""));
+        for (Entity entity : results.asIterable()) {
+            //Display x amount of entities taken from user input
+            String firstName = (String) entity.getProperty("firstname");
+            String lastName = (String) entity.getProperty("lastname");
+            String email = (String) entity.getProperty("email");
+            String subject = (String) entity.getProperty("subject");
+
+            tasks.add(new Task(firstName,lastName,email,subject));
+            
+            if (displayTasks == 0){
+                break;
+            }
+            else{
+                displayTasks--;
+            }
+        }
+
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(tasks));
+    }
+
+    private String convertToJsonUsingGson(List <Task> tasks) {
+        String json = gson.toJson(tasks);
+        return json;
   }
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        String value = request.getParameter(name);
+        if (value == null) {
+            return defaultValue;
+        }
+        return value;
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        long timestamp = System.currentTimeMillis();
+
+        // Get the input from the form.
+        String firstName = getParameter(request, "firstname", "");
```

Comment:
create public static final constants for these String parameter names. you can add javadoc to these constants to note that they're the POST request parameter keys as well as the Entity property keys